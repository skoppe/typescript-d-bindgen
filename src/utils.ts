import * as ir from './ir';

export function hasTypeGotHandle(type: ir.Type) : boolean {
    switch (type.type) {
        case 'intersection':
        case 'union':
            return false;
        case 'reference':
            const declaration = type.declaration();
            switch (declaration.declaration) {
                case 'struct': return true;
                case 'alias' : return hasTypeGotHandle(declaration.type)
                case 'enum' : return false;
            }
            throw new Error(`unknown declaration ${declaration.declaration}`)
        case 'array':
        case 'function': return false;
        case 'optional': return false;
        case 'handle': return false;
    }
    return false;
}

export enum FunctionKind {
    setter = 0,
    getter = 1,
    root = 2,
    nomangle = 3
}

export type Nominal<T, Name extends string> = T & { [Symbol.species]: Name; };

export type MangledName = Nominal<string, "MangledName">

export function mangleFunctionName(name: string, args: ir.Argument[], kind: FunctionKind) : MangledName {
    // TODO: for overloads we need to involve the args in the mangling
    function getPostfix() : string {
        switch (kind) {
            case FunctionKind.setter: return '_s';
            case FunctionKind.getter: return '_g';
            case FunctionKind.root: return '_r';
            case FunctionKind.nomangle: return '';
        }
        throw new Error(`Unable to get functionkind ${kind} postfix`)
    }
    return (name + getPostfix()) as MangledName;
}

export function mangleMethod(struct: ir.Struct, member: ir.StructMember, args: ir.Argument[], kind: FunctionKind) : MangledName {
    return mangleFunctionName(`${struct.name}_${member.name}`, args, kind)
}

export function isNotVoid(type: ir.Type) : boolean {
    return !(type.type === 'keyword' && type.name === 'void');
}
