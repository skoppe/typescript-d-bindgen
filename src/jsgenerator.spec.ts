import { getDeclarations } from './utils';
import * as jsgenerator from './jsgenerator'

function getDecoders(content: string) {
    const declarations = getDeclarations({
        name: 'filename.ts',
        content
    });
    const types = jsgenerator.extractTypesFromDeclarations(declarations);
    return jsgenerator.getDecoders(types.decoders);
}

test('union-string-int', () => {
    expect(getDecoders(`export declare function createChart(container: string | number): void;`))
        .toMatchSnapshot();
});

test('union-string-handle', () => {
    expect(getDecoders(`export declare function createChart(container: string | HTMLElement): void;`))
        .toMatchSnapshot();
});


test('enum-ints', () => {
    expect(getDecoders(`export declare const enum LineType { Simple = 0, WithSteps = 1 }; export declare function createLine(type : LineType): void;`))
        .toBe("");
});

test('enum-ints-no-init', () => {
    expect(getDecoders(`export declare const enum LineType { Simple, WithSteps }; export declare function createLine(type : LineType): void;`))
        .toBe("");
});

test('enum-strings', () => {
    expect(getDecoders(`export declare const enum LineType { s = "Simple", w = "WithSteps" }; export declare function createLine(type : LineType): void;`))
        .toBe("");
});

test('intersection', () => {
    expect(getDecoders(`export declare interface A { name: string }; export declare interface B { age: number }; export declare type Common = A & B; function fun(container: Common): void;`))
        .toBe("");
});

test('optional-string', () => {
    expect(getDecoders(`export declare function createChart(container: string | null): void;`))
        .toMatchSnapshot();
});

test('optional-double', () => {
    expect(getDecoders(`export declare function createChart(container: double | null): void;`))
        .toMatchSnapshot();
});

test('optional-interface', () => {
    expect(getDecoders(`export declare interface A { name: string }; function fun(container: Common | null): void;`))
        .toBe("");
});
// todo: optional templated interface
// todo: optional union
