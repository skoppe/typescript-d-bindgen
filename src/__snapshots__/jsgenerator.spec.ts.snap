// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`intersection 1`] = `""`;

exports[`union-string-handle 1`] = `
"const spasm_decode_union2_string_HTMLElement: (ptr) => {
    if (getUInt(ptr) === 0)
        return spasm_decode_string(ptr+4);
    if (getUInt(ptr) === 1)
        return spasm_decode_handle(ptr+4);
};"
`;

exports[`union-string-int 1`] = `
"const spasm_decode_union2_string_double: (ptr) => {
    if (getUInt(ptr) === 0)
        return spasm_decode_string(ptr+4);
    if (getUInt(ptr) === 1)
        return getDouble(ptr+4);
};"
`;
